WIP: Approval Voting

ziyan-junaideen opened the pull request:
## References

#2841
#2845 (the original PR)

## Objectives

Users should be able to select "approval" voting, as the voting style for each group. If approval voting is selected, then the user is constrained to a maximum number of votes per group.

## Visual Changes

This is a continuation of #2845, refer issue for more details. Changes in v0.18 have resulted changes the Budget Group form and is included below.

![Admin (1)](https://user-images.githubusercontent.com/762159/57306665-db9dc780-7100-11e9-9f55-0b93592fa6c6.png)
![Admin (2)](https://user-images.githubusercontent.com/762159/57306721-f6703c00-7100-11e9-82ab-3aa375f9ed47.png)
![Admin (3)](https://user-images.githubusercontent.com/762159/57306931-464f0300-7101-11e9-91e1-bda3eaef3b95.png)
![Participatory budgeting](https://user-images.githubusercontent.com/762159/57307458-3edc2980-7102-11e9-82c6-b48a64f98db0.png)
[with budget related restrictions on]
![Participatory budgeting (1)](https://user-images.githubusercontent.com/762159/57307560-6b904100-7102-11e9-9c8b-0c05beee39f9.png)
[voting after restriction lift]
![Participatory budgeting (2)](https://user-images.githubusercontent.com/762159/57307833-e0fc1180-7102-11e9-8f32-dd8b4bdce795.png)
[Sidebar vote details]
![CONSUL](https://user-images.githubusercontent.com/762159/57307617-85ca1f00-7102-11e9-8272-e3ed8be0eb64.png)


## Notes

In my list of todos:

- `number_votes_per_heading` was discussed to be renamed, but missed it
- complete tests

Devin from PBNYC told me that the core team would be interested in covering the tests and to make this PR.

ziyan-junaideen commented:
That is odd, Robocop in VSCode didn't highlight those issues. I can look in to it later.

voodoorai2000 commented on db/migrate/20180810211514_add_voting_style_and_number_votes_per_heading_to_groups.rb:
```diff
@@ -0,0 +1,6 @@
+class AddVotingStyleAndNumberVotesPerHeadingToGroups < ActiveRecord::Migration
+  def change
+    add_column :budget_groups, :voting_style, :string, default: 'knapsack'
```
Good setup for `voting_styles`, opening possibilities for other voting mechanisms 👌

voodoorai2000 replied:
For knapsack budgets that are currently in use, I think we would need a rake task to update the existing budgets 🤔, otherwise this field would still be nil.

ziyan-junaideen replied:
Sure, I will add a rake task.

ziyan-junaideen replied:
Apparently not...

I remember doing this some time back, but from recent I didn't as it seemed to do the trick. I had a dev seed db on v0.19 and then just ran the migration in this branch and it seems to have updated existing records.

```
consul_upstream=# select * from budget_groups where id = 1;
 id | budget_id | name |   slug   | max_votable_headings | voting_style | number_votes_per_heading 
----+-----------+------+----------+----------------------+--------------+--------------------------
  1 |         2 |      | all-city |                    1 | knapsack     |                        1
```

I already wrote the rake task, can include it in the PR if required.

voodoorai2000 replied:
Oh that's right, because it's a new attribute 👍Right on, no need for an additional rake task. 🎉

voodoorai2000 commented on app/models/budget/group.rb:
```diff
@@ -9,11 +9,17 @@ class Group < ActiveRecord::Base
     validates :budget_id, presence: true
     validates :name, presence: true, uniqueness: { scope: :budget }
     validates :slug, presence: true, format: /\A[a-z0-9\-_]+\z/
+    validates :voting_style, inclusion: { in: Vote::KINDS }
```
Good validations for `voting_style` and `number_of_votes_per_headings` 👌

voodoorai2000 commented on spec/features/budgets/votes_spec.rb:
```diff
@@ -173,4 +173,29 @@
   end
+
+
+  feature 'Approval Voting' do
```
It seems there is a related failing spec  https://travis-ci.org/consul/consul/jobs/482549751#L1887

ziyan-junaideen replied:
I think there were tests to be written. I only finished some of the comments of the previous PR.

ziyan-junaideen replied:
I will update the PR with the tests taken care of.

voodoorai2000 replied:
Thanks for the increased test coverage 👏

voodoorai2000 commented on app/models/budget/investment.rb:
```diff
@@ -233,6 +233,7 @@ def reason_for_not_being_ballotable_by(user, ballot)
       return :not_selected                    unless selected?
       return :no_ballots_allowed              unless budget.balloting?
       return :different_heading_assigned_html unless ballot.valid_heading?(heading)
+      return :not_enough_available_votes_html if heading.group.approval_voting? && ballot.investments.count == heading.group.number_votes_per_heading
```
How about moving these conditionals to a method?

voodoorai2000 replied:
Could you review the other @houndci-bot alerts? 🙏

ziyan-junaideen replied:
Yes of course. I remember doing some thing like that, may be another project.

voodoorai2000 commented on app/models/vote.rb:
```diff
@@ -1,2 +1,3 @@
 class Vote < ActsAsVotable::Vote
+  KINDS = %w(knapsack approval)
```
Why this distinction in the `Vote` class? Where is it used?

ziyan-junaideen replied:
@voodoorai2000 it is used in `Budget::Group` for validating the `voting_style` attribute. If I remember correct it doesn't have any functional use in the vote class. Shall I move it to `Budget::Group`?

voodoorai2000 replied:
+1 to moving it somewhere else. How about to the `Budget` class itself? 🤔

ziyan-junaideen replied:
I will move it there and make it `VOTING_STYLES = ...`

voodoorai2000 commented on app/views/budgets/ballot/_ballot.html.erb:
```diff
@@ -27,16 +27,19 @@
+                <span>
+                  <%= @ballot.formatted_amount_spent(@ballot.heading_for_group(group)) %>
+                </span>
+              <%end %>
```
A space before `end` :pray:

voodoorai2000 commented on app/views/budgets/ballot/_progress_bar.html.erb:
```diff
@@ -2,28 +2,40 @@
-                              @ballot.amount_spent(@heading)) %>">
-  </div>
-</div>
+<% if @heading.group.approval_voting? %>
```
How about using partials to display approval or knapsack content?

Something like this:
```
<% if approval_voting? %>
  <%= render "progress_bar_approval" %>
<% else %>
  <%= render "progress_bar" %>
<% end %>
```

ziyan-junaideen replied:
Sure.

voodoorai2000 commented on config/locales/val/budgets.yml:
```diff
@@ -4,9 +4,13 @@ val:
         title: Els meus vots
         amount_spent: Cost total
-        remaining: "Et queden <span>%{amount}</span> per a invertir."
+        remaining:
```
No need to add translations in other languages 😌 

Only English and Spanish are required, the other languages are managed through [Crowdin](https://crowdin.com/project/consul). 

We are still polishing the integration with Crowdin and these changes could create some conflicts 🤔 
Would you mind only adding the translations for English and Spanish? 🙏

ziyan-junaideen replied:
Sure. I will remove the rest.

voodoorai2000 commented on config/locales/es/budgets.yml:
```diff
@@ -39,6 +43,9 @@ es:
       finished: Resultados
+    voting_style:
+      knapsack: Bolsa de dinero
+      approval: Por aprovación
```
Typo: it should be "aprobación".

voodoorai2000 commented on config/locales/en/budgets.yml:
```diff
@@ -86,7 +93,9 @@ en:
             one: "<strong>You voted one proposal with a cost of %{amount_spent}</strong>"
             other: "<strong>You voted %{count} proposals with a cost of %{amount_spent}</strong>"
-          voted_info: You can %{link} at any time until the close of this phase. No need to spend all the money available.
+          voted_info:
```
Missing [view adaptation](https://github.com/consul/consul/blob/3d68d439a074abba389fc279bd1169f6567be5ea/app/views/budgets/ballot/_ballot.html.erb#L14) to new translation keys.

ziyan-junaideen replied:
Could you elaborate please. Do you mean this keys are not used?

voodoorai2000 replied:
Ah yes I think this is already solved, I was referring to updating this key in the view https://github.com/consul/consul/pull/3199/files#diff-8549275dea93f8da3dd8c13ffbcad0f7R18

voodoorai2000 commented on config/locales/es/budgets.yml:
```diff
@@ -151,6 +160,7 @@ es:
     progress_bar:
       assigned: "Has asignado: "
       available: "Presupuesto disponible: "
+      remaining_votes_html: "Has seleccionado <strong>%{casted_votes}</strong> projectos de <strong>%{votes_available}</strong>."
```
Typo: "projectos" should be "proyectos"

devinbalkind commented:
I can tackle these questions:

> Are users supposed to be able to vote only in one heading per group or multiple headings per group?

For our current use case they'd only vote in one but in a future use case they vote for more than one. Common would be voting in a city-wide process and a district/heading process.

> Do Approval Budgets still have a monetary value for every heading?

Yes - but it should be optional. Currently our users want to use that field to promote how much money they're giving away but from a system perspective it's inconsequential. 

> Do Approval Budget Investments get assigned a monetary value?

Yes they do. I can imagine use cases where they wouldn't but for now I think we should move forward as if they all will. 

If you have more questions please let me know.  :)

Thank you!

voodoorai2000 commented on app/controllers/admin/budget_groups_controller.rb:
```diff
@@ -11,7 +11,7 @@ def index
 
   def new
-    @group = @budget.groups.new
+    @group = @budget.groups.new(voting_cant_exceed_heading_budget: false)
```
Naming is always tricky, we try to keep names short, descriptive and positive 😌I wonder if there is a better name for this 🤔What would be a good word to describe being limited by money? 🤔 Maybe... `bounded`, `restricted` (as you use in specs), `money_bound`. Suggestions welcomed 😌

ziyan-junaideen replied:
I was thinking to use `restricted` but was worried that some other feature might find the name more appropriate. I am happy to any. I will use bounded since its the first thing in the list?

ziyan-junaideen replied:
I am going to use money_bound. It hints what it is.

voodoorai2000 commented on app/controllers/admin/budget_groups_controller.rb:
```diff
@@ -58,7 +58,11 @@ def groups_index
-      valid_attributes = [:max_votable_headings]
+      valid_attributes = [
+        :max_votable_headings,
+        :number_votes_per_heading,
```
How about `max_votes_per_heading`, I think it's a little more consistent with `max_votable_headings` or maybe `max_votes` if we move it to the `heading` table? 🤔

voodoorai2000 commented on app/helpers/ballots_helper.rb:
```diff
@@ -4,4 +4,12 @@ def progress_bar_width(amount_available, amount_spent)
     (amount_spent / amount_available.to_f * 100).to_s + "%"
   end
 
+  def remaining_votes(ballot, group)
```
Something is confusing here 🤔 We are looking at the number of votes per heading, but then looking at the votes in the group.

¿How about moving the `max_votes` to the `heading` class and comparing with the number of `votes` in the `heading`?

ziyan-junaideen replied:
Hmm...
What was I thinking...

ziyan-junaideen replied:
@voodoorai2000 Yah the code doesn't make sense. We can either fix this or as you suggested here, and hinted in a comment above move the max votes to the heading class. I am going to fix this for now. Let me know what you think is good.

ziyan-junaideen replied:
The remaining votes at the moment is:

          group.headings.count * group.number_votes_per_heading - ballot.investments.by_group(group.id).count

voodoorai2000 commented on app/helpers/budgets_helper.rb:
```diff
@@ -22,6 +22,10 @@ def budget_phases_select_options
   end
 
+  def budget_voting_styles_select_options
+    Budget::Vote::KINDS.map { |vk| [ t("budgets.voting_style.#{vk}"), vk] }
```
We try to use full names instead of acronyms, makes life happier 😌
¿How about using a variable name `kind` or `style`?

ziyan-junaideen replied:
Yah, Ruby generally doesn't use acronyms. I think `style` makes more sense. Will update.

voodoorai2000 commented on app/views/budgets/ballot/_progress_bar.html.erb:
```diff
@@ -2,28 +2,4 @@
-    </p>
-  </span>
-</div>
+<%= render "budgets/ballot/progress_bar/#{@budget.voting_style}" %>
```
Thank you for cleaning this up 👏

voodoorai2000 commented on config/locales/es/budgets.yml:
```diff
@@ -21,6 +25,7 @@ es:
         not_enough_money_html: "Ya has asignado el presupuesto disponible.<br><small>Recuerda que puedes %{change_ballot} en cualquier momento</small>"
         no_ballots_allowed: El periodo de votación está cerrado.
         different_heading_assigned_html: "Ya has votado proyectos de otra partida: %{heading_link}"
+        not_enough_available_votes_html: "No tines mas votos disponibles"
```
Typo: "No tienes", instead of "No tines".

voodoorai2000 commented on config/locales/fr/budgets.yml:
```diff
@@ -4,9 +4,12 @@ fr:
         title: Votre vote
         amount_spent: Montant dépensé
-        remaining: "Il vous reste <span>%{amount}</span> à investir."
+        remaining:
```
We only translate in English and Spanish in Github. 👌 The other translations we handle them through Crowdin, makes things simpler. Feel free to remove translations for other languages 👍

ziyan-junaideen replied:
Will remove...

voodoorai2000 commented on db/migrate/20190418114431_add_approval_voting_fields.rb:
```diff
@@ -0,0 +1,7 @@
+class AddApprovalVotingFields < ActiveRecord::Migration
```
Not sure about this DB structure: 🤔 

I think the simplest would be to add all approval voting attributes to the `budgets` table. But if we want to allow for different number of votes in different `headings` or `groups` it would make sense to move these attributes to the `groups` or `headings` tables. Taking into account your requirements in NY, would it make sense to keep the `voting_style` and `voting_cant_exceed_heading_budget` in the `budgets` table and the `number_votes_per_heading` in the `budget_headings` table?

ziyan-junaideen replied:
I pinged Devin, he is ok with it. I will move the fields.

ziyan-junaideen replied:
@voodoorai2000 : I pinged @devinbalkind on this. The 'exceed' flag should be moved to Budget. In the case of voting t would be useful to have the ability to mark the number of votes per heading in a per group basis.

ziyan-junaideen replied:
@voodoorai2000 : sorry I missed the last part. I think that will be good. I will ask devin and set it up.

voodoorai2000 commented on db/schema.rb:
```diff
@@ -35,8 +35,8 @@
     t.string   "action"
-    t.integer  "actionable_id"
     t.string   "actionable_type"
+    t.integer  "actionable_id"
```
Could you regenerate the `schema.rb` file with only the changes to the tables we are updating in this PR? 🙏

voodoorai2000 commented on lib/tasks/budgets.rake:
```diff
@@ -13,4 +13,11 @@ namespace :budgets do
 
   end
 
+  namespace :groups do
```
I think we don't need this rake task anymore as the `voting_style` is a new attribute and it would automatically add the default `knapsack` value to existing Budgets, as you pointed out previously 👍

voodoorai2000 commented on spec/factories/budgets.rb:
```diff
@@ -71,6 +79,10 @@
       association :budget, factory: [:budget, :drafting]
     end
+
+    trait :unrestricted do
```
Ideally we will find a name that can be used as both a trait and an attribute in DB. Maybe `restricted` is the one. 🤔

voodoorai2000 commented on spec/features/admin/budget_groups_spec.rb:
```diff
@@ -129,12 +129,12 @@
   context "Edit" do
 
     scenario "Show group information" do
+      budget.update_attribute :voting_style, "approval"
```
It's probably best to keep the approval and knapsack specs separate, to make sure we have covered all differences and that we are not changing the behaviour of the knapsack style. How about copying this spec to an "approval" `context` with the corresponding checks?

ziyan-junaideen replied:
This is some stray code.
And now the budget group is no longer modified.
Its fields moved to budget.

voodoorai2000 commented on spec/features/admin/budget_groups_spec.rb:
```diff
@@ -169,6 +171,23 @@
       expect(group.slug).to eq "new-english-name"
     end
 
+    scenario "Shows number of votes per heading if approval voting", :js do
```
This spec would look great inside a context for "approval voting" too 👌

voodoorai2000 commented on spec/features/budgets/investments_spec.rb:
```diff
@@ -1915,6 +1915,35 @@ def investments_order
       end
     end
+
+    feature "Voting Style" do
```
To keep spec names consistent how about something like this?:

```
context "Approval Voting" do
  context "Balloting Phase" do
    scenario "Add investment to ballot", :js do
```

voodoorai2000 commented on spec/features/budgets/votes_spec.rb:
```diff
@@ -1,9 +1,9 @@
 
-  feature "Investments" do
-    let(:manuela) { create(:user, verified_at: Time.current) }
+  feature "Investments - Knapsack" do
```
How about using a context here too? something like this:

```
context "Knapsack"
  feature "Investments" do
  end
end

context "Approval Voting"
  feature "Investments" do
  end
end
```

Note that the new Rails 5.1 syntax for `feature` is now `describe`. Feel free to use the new syntax 🎉
https://github.com/consul/consul/pull/3562/commits/307cf248468102b9afef0efab7fd8d387601d88a

ziyan-junaideen replied:
I will try it. I thought Capybara used `feature` and `scenario`.

voodoorai2000 commented on spec/helpers/ballots_helper_spec.rb:
```diff
@@ -0,0 +1,28 @@
+
+describe BallotsHelper do
+
+  describe "#remaining_votes" do
```
Thank you for using helper specs, they are great and super fast to run 👍

No strong preference, but I think you could use `let!(:line)`, that should avoid the need to run the `before` block, as it ejecutes the initialization right away instead of waiting to the variable to be called 😉

voodoorai2000 commented on spec/models/budget/ballot/line_spec.rb:
```diff
@@ -24,6 +24,13 @@
+      it "is valid if insufficient funds, is approval voting, limits turned off" do
+        budget.update_attribute :voting_style, "approval"
+        group.update_attribute :voting_cant_exceed_heading_budget, false
+        investment.update_attribute :price, investment.price + 1
```
Shouldn't this be `heading.price + 1` instead of `investment.price + 1`? 🤔

ziyan-junaideen replied:
Good question. Checking.

decabeza commented on app/views/budgets/ballot/_ballot.html.erb:
```diff
@@ -26,16 +32,19 @@
-              <%= @ballot.formatted_amount_spent(@ballot.heading_for_group(group)) %>
-            </span>
+              <% if @budget.approval_voting? %>
+                <%= t('budgets.ballots.show.votes_cast_html', amount: @ballot.investments.by_group(group.id).count) %>
```
Please, could you change it to double quotes? 🙏😌

decabeza commented on app/views/budgets/ballot/progress_bar/_approval.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="panel clear">
+  <p>
+    <%= t('budgets.progress_bar.remaining_votes_html',
```
Please, could you change it to double quotes? 🙏😌

decabeza commented on app/views/budgets/investments/_sidebar.html.erb:
```diff
@@ -15,10 +15,9 @@
-  </p>
+    <%= t("budgets.investments.index.sidebar.voted_info.#{@budget.voting_style}",
+          link: link_to(t("budgets.investments.index.sidebar.voted_info_link"), budget_ballot_path(@budget))) %>
+ p>
```
This tag should be `</p>` 😉

ziyan-junaideen commented on spec/features/admin/budgets_spec.rb:
```diff
@@ -126,6 +126,33 @@
+      expect(budget.knapsack_voting?).to be true
+    end
+
+    scenario "Create budget - voting style approval", :js do
```
@voodoorai2000 Is this ok, or will it be good to split it in to some contexts.

```
context "New" do
  context "Approval" do
    scenario "Create budget"
  end

  context "Knapsack" do
    scenario "Create budget"
  end

  // ...
end
```
I have flattened it out because the other budget actions are not effected by if it is approval or knapsack.

```
context "New" do
  scenario "Create budget - Knapsack"
  scenario "Create budget - Approval"
  // ...
end
```

voodoorai2000 replied:
Sure that's fine 👍

ziyan-junaideen commented:
@voodoorai2000 : i think i have addressed your comments. At the moment specs are running for a failing spec which was a result of not enabling JS.

ziyan-junaideen commented:
The PR had a failing spec earlier. It was related to actions. When I checked locally the spec passed. When I restarted the Travis job then it passed. This is not the first time I have seen a test fail in a non deterministic fashion. Its not limited to this PR. I also saw it in the other feature I was working on.

```
  1) Budget Investments sidebar map Approval Voting Balloting Phase Add an investment to ballot
     Failure/Error: find(".add a").click
     
     ActionController::UnknownFormat:
       Budgets::Ballot::LinesController#create is missing a template for this request format and variant.
     
       request.formats: ["text/html"]
       request.variant: []
     # ./spec/support/common_actions/budgets.rb:9:in `block in add_to_ballot'
     # ./spec/support/common_actions/budgets.rb:8:in `add_to_ballot'
     # ./spec/features/budgets/investments_spec.rb:1981:in `block (5 levels) in <top (required)>'
```

ziyan-junaideen commented:
@voodoorai2000 : a little advice if you have seen some thing similar. 

I am having a odd issue. Travis is showing one error, I am experiencing another error locally.

Local error: https://gist.github.com/ziyan-junaideen/a040ac1a9f09b99741783fefdd63643f
Travis: https://travis-ci.org/ziyan-junaideen/consul/jobs/569165568#L1164

In summary...

**Locally** i get stuck after clicking on 'Collaborative Legislation' and then it doesn't read the 'All' tab in the filters section. How ever if the URL is loaded directly it will run. I added some logging and noticed the execution going through both the controller and view. I changed the JS driver to `chrome` in the placed a break-points and noticed the execution going through the controller and view. I wonder if this is a Turbolinks issue like loading the page async?

ziyan-junaideen commented:
@voodoorai2000  @decabeza Tests are passing
But the coverage is not
When I checked it seems to highlight other controller/models?

javierm commented:
@ziyan-junaideen Thanks a lot! :pray:

As agreed, I've openen pull request #4063, updating this pull request to our latest code :smile:.

Closing in favor of #4063.
